api/opensource/mfx_dispatch/linux/mfxloader.cpp

```diff
@@ -0,0 +1,582 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <assert.h>
+#include <dlfcn.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <algorithm>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include "mfxvideo.h"
+#include "mfxplugin.h"
+#include "mfxpak.h"
+
+#include "mfxloader.h"
+
+namespace MFX {
+
+#if defined(__i386__)
+#define LIBMFXSW "libmfxsw32.so.1"
+#define LIBMFXHW "libmfxhw32.so.1"
+#elif defined(__x86_64__)
+#define LIBMFXSW "libmfxsw64.so.1"
+#define LIBMFXHW "libmfxhw64.so.1"
+#else
+#error Unsupported architecture
+#endif
+
+#undef FUNCTION
+#define FUNCTION(return_value, func_name, formal_param_list, actual_param_list) \
+  e##func_name,
+
+enum Function
+{
+  eMFXInit,
+  eMFXInitEx,
+  eMFXClose,
+  eMFXJoinSession,
+#include "mfxvideo_functions.h"
+  eFunctionsNum,
+  eNoMoreFunctions = eFunctionsNum
+};
+
+struct FunctionsTable
+{
+  Function id;
+  const char* name;
+  mfxVersion version;
+};
+
+#define VERSION(major, minor) {{minor, major}}
+
+#undef FUNCTION
+#define FUNCTION(return_value, func_name, formal_param_list, actual_param_list) \
+    { e##func_name, #func_name, API_VERSION },
+
+static const FunctionsTable g_mfxFuncTable[] =
+{
+    { eMFXInit, "MFXInit", VERSION(1, 0) },
+    { eMFXInitEx, "MFXInitEx", VERSION(1, 14) },
+    { eMFXClose, "MFXClose", VERSION(1, 0) },
+    { eMFXJoinSession, "MFXJoinSession", VERSION(1, 1) },
+#include "mfxvideo_functions.h"
+    { eNoMoreFunctions }
+};
+
+typedef mfxStatus (MFX_CDECL *CreatePluginPtr)(mfxPluginUID, mfxPlugin*);
+
+class LoaderCtx;
+
+class PluginCtx
+{
+public:
+  PluginCtx(LoaderCtx& loader)
+    : m_loader(loader)
+  {}
+
+  mfxStatus Load(const mfxPluginUID& uid, mfxU32 version, const char *path);
+  mfxStatus Unload();
+
+  inline mfxPluginUID getUID() { return m_uid; }
+
+private:
+  LoaderCtx& m_loader;
+  std::shared_ptr<void> m_dlh;
+  CreatePluginPtr m_create_plugin = nullptr;
+  mfxPluginUID m_uid{};
+  mfxPlugin m_plugin{};
+  mfxPluginParam m_plugin_param{};
+};
+
+class LoaderCtx
+{
+public:
+  mfxStatus Init(mfxInitParam par);
+  mfxStatus Close();
+
+  mfxStatus LoadPlugin(const mfxPluginUID& uid, mfxU32 version, const char *path);
+  mfxStatus UnloadPlugin(const mfxPluginUID& uid);
+
+  inline void* getFunction(Function func) {
+    return m_table[func];
+  }
+
+  inline mfxSession getSession() {
+    return m_session;
+  }
+
+  inline mfxIMPL getImpl() {
+    return m_implementation;
+  }
+
+  inline mfxVersion getVersion() {
+    return m_version;
+  }
+
+private:
+  std::shared_ptr<void> m_dlh;
```

Comment:
You know, as of now it is not really shared. However, it is done via std::shared_ptr on purpose (I considered and tried std::unique_ptr) because it is more difficult to wrap it in std::unique_ptr because of the difference in their implementations. Essentially, you will be forced to declare:
`typedef std::unique_ptr<void,  int (*)(void *)> unique_dlopen_ptr;`
and make other changes which are simply hidden by std::shared_ptr. In other words, std::shared_ptr seems more easy to use for this particular purpose to smart-wrap dlopen.